[SPARK-6990] [Build] Add Java linting script; fix minor warnings

dskrvk opened the pull request:
This replaces https://github.com/apache/spark/pull/9696

Invoke Checkstyle and print any errors to the console, failing the step.
Use Google's style rules modified according to
https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide
Some important checks are disabled (see TODOs in `checkstyle.xml`) due to
multiple violations being present in the codebase.

Suggest fixing those TODOs in a separate PR(s).

More on Checkstyle can be found on the [official website](http://checkstyle.sourceforge.net/).

Sample output (from [build 46345](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46345/consoleFull)) (duplicated because I run the build twice with different profiles):

> Checkstyle checks failed at following occurrences:
> [ERROR] src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:[217,7](coding) MissingSwitchDefault: switch without "default" clause.
> [ERROR] src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java:[198,10](modifier) ModifierOrder: 'protected' modifier out of order with the JLS suggestions.
> [ERROR] src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:[217,7](coding) MissingSwitchDefault: switch without "default" clause.
> [ERROR] src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java:[198,10](modifier) ModifierOrder: 'protected' modifier out of order with the JLS suggestions.
> [error] running /home/jenkins/workspace/SparkPullRequestBuilder@2/dev/lint-java ; received return code 1

Also fix some of the minor violations that didn't require sweeping changes.

Apologies for the previous botched PRs - I finally figured out the issue.

cr: @JoshRosen, @pwendell 

> I state that the contribution is my original work, and I license the work to the project under the project's open source license.

dskrvk committed:
[SPARK-6990] Add Java linting script

Invoke Checkstyle and print any errors to the console, failing the step.
Use Google's style rules modified according to
https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide
Some important checks are disabled (see TODOs in checkstyle.xml) due to
multiple violations being present in the codebase.

dskrvk committed:
[SPARK-6990] Fix some Checkstyle warnings

dskrvk committed:
[SPARK-6990] Fix some Checkstyle warnings

dskrvk committed:
[SPARK-6990] Fix some Checkstyle warnings

dskrvk committed:
[SPARK-6990] Fix some Checkstyle warnings

dskrvk committed:
[SPARK-6990] Suppress Checkstyle for TimSort

The code was copied from a third-party source and needs to be in sync with
that, so we shouldn't make our own modifications to it.
The file contains some style violations, so suppress the checks.

dskrvk committed:
[SPARK-6990] Suppress Checkstyle for TimSort

The code was copied from a third-party source and needs to be in sync with
that, so we shouldn't make our own modifications to it.
The file contains some style violations, so suppress the checks.

dskrvk committed:
[SPARK-6990] Suppress Checkstyle for TimSort

The code was copied from a third-party source and needs to be in sync with
that, so we shouldn't make our own modifications to it.
The file contains some style violations, so suppress the checks.

dskrvk committed:
[SPARK-6990] Suppress Checkstyle for TimSort

The code was copied from a third-party source and needs to be in sync with
that, so we shouldn't make our own modifications to it.
The file contains some style violations, so suppress the checks.

dskrvk committed:
[SPARK-6990] Fix some Checkstyle issues in tests

dskrvk committed:
[SPARK-6990] Fix some Checkstyle issues in tests

dskrvk committed:
[SPARK-6990] Enable Checkstyle for tests

dskrvk committed:
[SPARK-6990] Enable Checkstyle for tests

dskrvk committed:
[SPARK-6990] Enable FallThrough check in Checkstyle

This makes sure all case statements end with a break. See
http://checkstyle.sourceforge.net/config_coding.html#FallThrough

dskrvk committed:
[SPARK-6990] Enable FallThrough check in Checkstyle

This makes sure all case statements end with a break. See
http://checkstyle.sourceforge.net/config_coding.html#FallThrough

dskrvk committed:
[SPARK-6990] Remove whitespace after '>'

Closing '>' and method name shouldn't have whitespace between them, according
to http://checkstyle.sourceforge.net/config_whitespace.html#GenericWhitespace

dskrvk committed:
[SPARK-6990] Disable MissingSwitchDefault check

Checks fails in UnsafeRowParquetRecordReader.java. Let's enable the check in a
separate change.

dskrvk committed:
[SPARK-6990] Disable MissingSwitchDefault check

Checks fails in UnsafeRowParquetRecordReader.java. Let's enable the check in a
separate change.

dskrvk committed:
[SPARK-6990] Fix qualifier order in SpecificParquetRecordReaderBase

dskrvk committed:
[SPARK-6990] Fix qualifier order in SpecificParquetRecordReaderBase

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

AmplabJenkins commented:
Can one of the admins verify this patch?

JoshRosen commented:
Jenkins, this is ok to test.

SparkQA commented:
**[Test build #46433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46433/consoleFull)** for PR 9867 at commit [`fd6d0e0`](https://github.com/apache/spark/commit/fd6d0e0a640b76c9f95f25fb2ab756b645efc69f).

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
**[Test build #46433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46433/consoleFull)** for PR 9867 at commit [`fd6d0e0`](https://github.com/apache/spark/commit/fd6d0e0a640b76c9f95f25fb2ab756b645efc69f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public class JavaLDAExample `\n  * `  public abstract static class PrefixComputer `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46433/
Test PASSed.

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge branch 'master' into spark-6990

dskrvk committed:
[SPARK-6990] Fix Checkstyle issues in ML examples

SparkQA commented:
**[Test build #46482 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46482/consoleFull)** for PR 9867 at commit [`7a49ad7`](https://github.com/apache/spark/commit/7a49ad708d35bfd1ba7027bb07ae016b2057987c).

SparkQA commented:
**[Test build #46482 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46482/consoleFull)** for PR 9867 at commit [`7a49ad7`](https://github.com/apache/spark/commit/7a49ad708d35bfd1ba7027bb07ae016b2057987c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  public abstract static class PrefixComputer `\n  * `abstract class Aggregator[-I, B, O] extends Serializable `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46482/
Test PASSed.

dskrvk commented:
Added some more commits so that new changes are in line with the style guide.

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk commented:
Ping? I've just merged the latest changes locally and verified the checks still pass.

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
**[Test build #46692 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46692/consoleFull)** for PR 9867 at commit [`c773e90`](https://github.com/apache/spark/commit/c773e90c93ce211f3bd348eb87fade2a2c19a025).

rxin commented:
I haven't looked at it super closely yet but I think this is definitely good to have!

SparkQA commented:
**[Test build #46692 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46692/consoleFull)** for PR 9867 at commit [`c773e90`](https://github.com/apache/spark/commit/c773e90c93ce211f3bd348eb87fade2a2c19a025).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  public abstract static class PrefixComputer `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46692/
Test PASSed.

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk commented:
Thanks @rxin. Would appreciate a "Ship it" on this (unless there are issues). Don't mean to whine, but the longer we wait, the bigger this PR becomes as I have to fix any new code that doesn't pass the checks. Right now the latest merge from upstream passes successfully.

Added some more details in the description.

JoshRosen commented:
Jenkins, retest this please.

JoshRosen commented:
Hey @dskrvk, sorry to let this slip through the cracks. I'm going to shepherd this today to try to get it merged.

SparkQA commented:
**[Test build #47074 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47074/consoleFull)** for PR 9867 at commit [`c773e90`](https://github.com/apache/spark/commit/c773e90c93ce211f3bd348eb87fade2a2c19a025).

JoshRosen commented on dev/lint-java:
```diff
@@ -0,0 +1,34 @@
+SPARK_ROOT_DIR="$(dirname $SCRIPT_DIR)"
+
+$SCRIPT_DIR/../build/mvn -Pkinesis-asl -Phive -Phive-thriftserver checkstyle:check > checkstyle.txt
+$SCRIPT_DIR/../build/mvn -Pkinesis-asl -Pyarn -Phadoop-2.2 checkstyle:check >> checkstyle.txt
```
Quick question: why do you need to run twice with different profiles? AFAIK the set of source files should be the same under all of the Hadoop profiles, so I don't think we need to set `-Phadoop-2.2` here.

Why can't we just use one Maven run with the profiles `-Pkinesis-asl -Phive -Phive-thriftserver -Pyarn`?

dskrvk replied:
I didn't actually observe any differences between the two profiles in terms of Checkstyle warnings, but decided to add the second run just to be thorough.

My reasoning was that since some profiles omit some of the modules, we need to exercise all of the possible ones, even though at the moment the set of Java sources may be the same. In any case, this only adds a few seconds to the build - negligible compared to the overall `run-tests` time.

vanzin replied:
You should be able to have a single run with "-Pkinesis-asl -Pyarn -Phive -Phive-thriftserver" - I even think "-Phive" is unnecessary, I think it only affects packaging right now.

"-Phadoop2.2" is unnecessary, that's the default.

dskrvk replied:
Good point; changed.

JoshRosen commented:
Changes look good to me and ready to merge today; my [only question](https://github.com/apache/spark/pull/9867#discussion_r46461693) concerns why we need to run Checkstyle twice with different sets of profiles.

SparkQA commented:
**[Test build #47074 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47074/consoleFull)** for PR 9867 at commit [`c773e90`](https://github.com/apache/spark/commit/c773e90c93ce211f3bd348eb87fade2a2c19a025).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  public abstract static class PrefixComputer `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47074/
Test PASSed.

vanzin commented:
Minor, but there's a typo in the title: "mix" -> "fix".

dskrvk committed:
Merge remote-tracking branch 'upstream/master'

dskrvk committed:
[SPARK-6990] Remove second run of mvn checkstyle

Add -Pyarn to the first run so that all modules are covered at once. No need to
execute checkstyle twice any more.

SparkQA commented:
**[Test build #47183 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47183/consoleFull)** for PR 9867 at commit [`b079f29`](https://github.com/apache/spark/commit/b079f29f5370a3fe1ce643ff2ab1c999da1d72b9).

SparkQA commented:
**[Test build #47183 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47183/consoleFull)** for PR 9867 at commit [`b079f29`](https://github.com/apache/spark/commit/b079f29f5370a3fe1ce643ff2ab1c999da1d72b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  public abstract static class PrefixComputer `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47183/
Test FAILed.

dskrvk commented:
Jenkins, retest this please.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #47201 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47201/consoleFull)** for PR 9867 at commit [`b079f29`](https://github.com/apache/spark/commit/b079f29f5370a3fe1ce643ff2ab1c999da1d72b9).

SparkQA commented:
**[Test build #47201 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47201/consoleFull)** for PR 9867 at commit [`b079f29`](https://github.com/apache/spark/commit/b079f29f5370a3fe1ce643ff2ab1c999da1d72b9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  public abstract static class PrefixComputer `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47201/
Test PASSed.

JoshRosen commented:
LGTM, so I'm going to merge this into master. Thanks for being so patient, @dskrvk!

JoshRosen commented:
Hey @dskrvk, what's your Apache JIRA username? I need it in order to assign the JIRA to you so that you are properly credit by our release-notes generation script.

dskrvk commented:
Hey @JoshRosen, thanks, my handle is dskrvk there as well.
